### ISSUE TITLE
Reimporting a mesh doesn't update the mesh in the Scene

### ISSUE THREAD

**Issue by Tisten** [OurMachinery/themachinery-public#1132]
- **Version:** Git Master Branch
- **OS:** Windows

Repro project:
[reimport_mesh.zip](https://github.com/OurMachinery/themachinery-public/files/8198906/reimport_mesh.zip)

* When reimporting a dcc_asset with a mesh, the preview tab updates as expected, but the entity using the old mesh doesn't get updated in the Scene tab.
* If extracting the assets from the reimported dcc_asset, the newly created entity looks correct
* If opening the old asset, the scene view is not totally empty

Video showing this:
https://user-images.githubusercontent.com/6726075/157069180-36c40438-eda6-4c07-a83a-260c6be497c9.mp4

**Comment by karl-zylinski-ourmachinery** [comment]
I have looked a bit at this now and while there is room for improvement in how we currently do things. However, I see some issues in the repro case:

- Both the scene tree node names and mesh names have changed.
  - Node change: `containerframe_01_msh` -> `containerframe_01_OBJ`.
  - Mesh change: `containerframe_01_msh:mockup_atlas_hardsurface_01_mat` -> `containerframe_01_MSH:mockup_atlas_hardsurface_01_mat`

The question is what the best way to handle this is. I have two proposals which may both be useful, but I'd like to hear your thoughts:

1. We supply a "re-rig" checkbox when doing reimports. This locates any entity rigged from the dcc asset and deletes all the meshes in there and re-rigs it from the new data.
2. We provide an interactive reimport tool where it figures out what meshes and nodes are missing in the newly imported asset vs the previously rigged entity and lets you remap them. It could also let you choose if previously unknown (i.e. new meshes) should be added.

1) Might be quite quick to make, but is it useful, or maybe too dangerous? Any changes you've made to the stuff under the `Meshes` entity in the rigged entity would disappear.

**Comment by Tisten** [comment]
Summoning @LoffanElemental as well, the artist who is trying to make this work.

We have discussed two ideas for this area:
1. Very similar to your number 1, throw away all entities created by the rigging. To reduce the dangerousness of it we would love to be able to mark the rigged (autogenerated) entities as "read only", i.e you can only override the non-generated parts of the entity like the root entity. This means the artist can trust that no one who uses the entity will suffer from the changes.
2. It would be awesome to replace the rigging code with a rigging graph tool, i.e a creation graph which creates an entity tree from dcc assets. In this case the technical artist doing these graphs can choose to either do like 1. and throw away the old entity and create a new one, or to parse the old entity and patch up the new one based on the old one. The complexity is in the hand of the creative guys, not the engine writers.

Just doing your number 1 is probably an okay start. I think that is what Loffan does right now, but manually. But as you say, using it comes with the risk of damaging other people's work and it is complicated/time consuming to find out what you will damage, if anything.

This is my view as a programmer, but I should not decide.

**Comment by LoffanElemental** [comment]
I think number one sounds like a good option for us! The way we want to work is that one fbx only holds one mesh. So the only time there are multiple meshes in one dcc asset is if there is an export error. But I agree with Tisten that it would be nice to have a way to control this if your setup is to use multiple meshes in one dcc asset. I'm a bit uncertain if nr 1 is problematic for Andy, our tech animation guy but for me it sounds good.

**Comment by ElementalAndy** [comment]
Nope, I'm all aboard. @karl-zylinski-ourmachinery is #1 conceptually the same as "merge images" and "merge materials"? Is the "merge entities" a better name?

I also think it would be very useful to have #2 in the future.

**Comment by karl-zylinski-ourmachinery** [comment]
I have now pushed a first pass of option (1). There is a checkbox in the import settings to rerig entities on reimport (checked by default). After reimporting a DCC asset it will find entities rigged from it and clean out the meshes and after that rerig the DCC asset into it. Try it and let me know how it works for you.

**Comment by LoffanElemental** [comment]
I have tried the feature. It imports the new model but it loses all settings like materials etc.
![image](https://user-images.githubusercontent.com/94915027/157843807-619d6ee0-31db-4cfc-9183-dc0deebe6f44.png)

**Comment by Tisten** [comment]
To clarify (both to Loffan and Karl), the Model Name is added by us, so that is custom data. And since everything created by the last rigging is thrown away and created fresh, it is expected that the material is reset to a default one. And I guess it wouldn't help to set these things as overrides since the UUIDs which identifies the entities and components are changed when recreated.

**Comment by LoffanElemental** [comment]
It is probably my lack of understanding of the tech behind the machinery. I just assumed that the data inputs could be saved and written to the new rigged mesh as a last step.

What I'm used to from other engines is that you press reimport and all set data is still there when the reimport is done. At Avalanche the engine even listened for mesh updates and automatically updated the mesh with all data intact.

**Comment by LoffanElemental** [comment]
But the first step is an improvement and I use reimport all the time now! So good work on the first step!

**Comment by karl-zylinski-ourmachinery** [comment]
I have reworked this now. The behavior is now less destructive. It updates the DCC asset but keeps the mapping to any old extracted assets. It extracts any previously unseen images and materials. It finds rigged entities and removes, updates and adds new meshes. The update is non-destructive, compared to before.

More info in commit and commit message: https://github.com/OurMachinery/themachinery/commit/16b04c90bbeaa9ebe0ca3902caf6c9097f32390b

Let me know if this solves this issue. @Tisten @LoffanElemental

**Comment by LoffanElemental** [comment]
The new stuff appears to work. I think there are to many choices in the extract menu right now. When I first tried I had "Re-extract" and "Update Rigged Entities" checked. That created 2 meshes and it lost all connections.
![image](https://user-images.githubusercontent.com/94915027/159706718-570dbfd1-f41b-4e19-b00b-622c27dbd57b.png)

But when I only use "Update Rigged Entities" It seams to work fine! Do we need the Re-extract on Reimport option now?

**Comment by LoffanElemental** [comment]
I ran into an issue today. I imported a mesh that had nonmanifold errors. This created two meshes in the entity when extracted. One with correct geometry and one with the nonmanifold geometry. So far so good. 
![image](https://user-images.githubusercontent.com/94915027/159957982-96e55b48-28d7-4b14-8ba9-e1b711a341ae.png)

But when I fixed the error in blender and pressed Reimport with "Update Rigged Entities on Reimport" I still get two meshes in the entity but they seam to be identical. 
![image](https://user-images.githubusercontent.com/94915027/159959006-b4354b22-f0f2-440c-8a34-8f02c4a6d541.png)

What I would expect is that the entity is updated to only hold one mesh to mirror the DCC_asset. 

I also tried using "Re-extract on Reimport" but it made no difference.

**Comment by LoffanElemental** [comment]
Other times when I try to reimport after fixing nonmanifold geometry the meshes get totally broken in the entity: 
![image](https://user-images.githubusercontent.com/94915027/159961285-46d38d28-a1cf-4136-a47e-5682ba908fd4.png)

The DCC asset always seams to reflect the correct things in the preview

**Comment by karl-zylinski-ourmachinery** [comment]
> The new stuff appears to work. I think there are to many choices in the extract menu right now.

Agreed. We are experimenting with maybe replacing the DCC asset import pipeline with a graph-based solution. So for now it has to be a bit chaotic until we get somewhere with that or some better solution.

> When I first tried I had "Re-extract" and "Update Rigged Entities" checked. That created 2 meshes and it lost all connections. ![image](https://user-images.githubusercontent.com/94915027/159706718-570dbfd1-f41b-4e19-b00b-622c27dbd57b.png)
> 
> But when I only used "Update Rigged Entities" It seams to work fine! Do we need the Re-extract on Reimport option now?

Strange! Re-extract is meant to pull out materials and images out of the DCC asset in case any new such have appeared. Could you supply some tiny repro project where this happens? And also something for the mesh with nonmanifold errors?

**Comment by Tisten** [comment]
Here is a recording of the import of the mesh with a manifold issue, and the reimport of the fixed one:
![Reimport](https://user-images.githubusercontent.com/6726075/160592901-b685ab39-6537-428d-90db-74a880d899ad.gif)

And here are the meshes so you can try for yourself:
[meshes for repro.zip](https://github.com/OurMachinery/themachinery-public/files/8370499/meshes.for.repro.zip)

The issue is that two meshes are left, even though the fixed model only have one mesh.

If you first import the fixed model, and then reimport it with the broken model instead, then only the first mesh (the broken one) will appear in the mesh after the reimport.

**Comment by karl-zylinski-ourmachinery** [comment]
I see what's wrong. The way we match up the meshes when reimporting is by comparing the name of the scene tree node and material (hashed together). But both these end up at the same node, so even if it only contains one mesh after reimport, it still thinks that both are valid since they both use a valid node-material-name. Do you usually have multiple meshes under a single node? I didn't think any DCC tool supported this?

**Comment by karl-zylinski-ourmachinery** [comment]
I re-read the comment above:
> I ran into an issue today. I imported a mesh that had nonmanifold errors. This created two meshes in the entity when extracted. One with correct geometry and one with the nonmanifold geometry.

Did the second mesh appear as an error in the import pipeline, i.e. does both the bad and fixed fbx contain only a single mesh?

**Comment by karl-zylinski-ourmachinery** [comment]
Nevermind, I looked a bit at the fbx data and looking at how the non manifold stuff works. I'll investigate this a bit more and see if I can come up with a solution.

**Comment by LoffanElemental** [comment]
When I use "Update Rigged Entities on Reimport" on the in the latest version I'm back to it whiping all saved info again :(
![image](https://user-images.githubusercontent.com/94915027/157843807-619d6ee0-31db-4cfc-9183-dc0deebe6f44.png)

**Comment by karl-zylinski-ourmachinery** [comment]
Update on the issues with the non-manifold stuff: This is an issue with assimp, the library we use for import of the fbx. When it runs into this kind of geometry it spawns two separate meshes instead of one, but both have the same name, so we have trouble differentiating between them upon reimport. Trying to figure something out. Meanwhile, you can safely manually delete the subentity that represents the mesh that is "broken".

**Comment by karl-zylinski-ourmachinery** [comment]
Hello. Sorry that I never was able to figure out the non-manifold issues. I ended up debugging assimp code and it didn't really get me anywhere. Down the line we might use our own GLTF importer and then maybe have a non-assimp FBX importer which we have more control over.

Are any of the other issues in this thread still an issue? @Tisten @LoffanElemental
